src/txdb.cpp

```diff
@@ -158,14 +195,39 @@ void BatchWriteAnchors(CDBBatch& batch, Map& mapToUse, const char& dbChar)
     }
 }
 
+void BatchWriteMMR(CDBBatch& batch, CHistoryCacheMap& historyCacheMap) {
+
+    for (auto nextHistoryCache = historyCacheMap.begin(); nextHistoryCache != historyCacheMap.end(); nextHistoryCache++) {
+        auto historyCache = nextHistoryCache->second;
+        auto epochId = nextHistoryCache->first;
+
+        // delete old entries since updateDepth
```

Comment:
I just noticed that this only stores a single copy of each epoch's tree corresponding to the state as of the last block in the epoch (or the current chain tip for the latest epoch). This is sufficient for the full node's internal usage, but means that any future RPC that wants to operate with the trees for past blocks would need to compute its trees from scratch. This does mean more work to implement support in e.g. lightwalletd, but it isn't necessarily a problem; there isn't a clear benefit to having the tree state for each block, because light clients will probably want trees for "recent intervals of blocks".

Reply:
You just take current tree, do several pops from it to the desired depth and don't commit it anywhere. Pop operation is logarithmic. To support forks, you pop to the fork block, then add headers from branch, each operation logarithmic as well.